Add post outlining networking initiative

simonrw opened the pull request:
The post in this PR announces and supports the networking initiative:
* giving an outline of what the initiative is trying to do,
* outlining the upcoming posts
* introducing the DNS server migration
* giving examples of user configuration

simonrw committed:
Start networking post

simonrw committed:
Add dig output example

simonrw committed:
Update main post and split out gateway listen content

simonrw committed:
Include initial sketch for logo

simonrw committed:
Mention debug tool

simonrw committed:
Rework introduction section

simonrw committed:
Update post

simonrw committed:
Remove next post draft

simonrw committed:
Remove draft status

So the post shows up in the branch preview

dfangl commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,173 @@
+
+LocalStack normally runs in a container, meaning that it is isolated from the host system.
+By default, LocalStack _publishes_ its edge port (usually 4566) to the host.
+Publishing a port means that a ports on the host forwards network communications to the LocalStack container.
```
```suggestion
Publishing a port means that a port on the host forwards network communications to the LocalStack container.
```

dfangl commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,173 @@
+
+This works well when interacting from the host, for example using `awslocal` commands.
+It does not work when trying to connect to LocalStack from your own containers, or LocalStack compute resources such as Lambda functions or ECS containers.
+In the past, some people have suggested connecting their application containers to the host network (`--network host`) or by making requests to `host.docker.internal:4566`.
```
We could mention the per default non-existence of host.docker.internal as a problem in the list below it but it might not be worth it. Just a random thought, because I somehow expected it to, but maybe it would be too much anyway.

simonrw replied:
I think that mentioning that `host.docker.internal` is not always present increases the validity of the workaround, rather than emphasising the new features, but thanks for raising!

simonrw committed:
Mention the demo application

simonrw committed:
Rephrase confusing sentence

dfangl commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,173 @@
+In the past, some people have suggested connecting their application containers to the host network (`--network host`) or by making requests to `host.docker.internal:4566`.
+In some cases, using the host networking solves the problem, but it raises other ones:
+
+* If SSL is used, then certificate validation must be turned off since LocalStack does not present a valid certificate for the domain used(either `localhost` or `host.docker.internal`).
```
I think this sentence might be conceived that the `(either localhost or host.docker.internal)` are the valid names for the certificate? Also, the `localhost` certificate problem is the same when connecting from the host to the container when using a bridge network, something we might (or not, if it gets too long) want to mention above as well.

dfangl replied:
Also, a space should be there before the parenthesis.

dfangl commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,173 @@
+
+## How to use this new feature
+
+For AWS services like Lambda or ECS, we are running your application code in a pre-configured environment.
```
Pre configured for what?
```suggestion
For AWS services like Lambda or ECS, we are running your application code in an environment pre-configured to use this feature.
```

dfangl commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
````diff
@@ -0,0 +1,173 @@
+```
+
+We hope that with this new functionality available today, accessing LocalStack should be considerably easier.
+By moving the DNS server into LocalStack and configuring AWS compute environments, your Lambda functions, ECS containers, and EC2 instances should already be able to access LocalStack at `localhost.localstack.cloud`.
````
Here I also find the usage of "configuring" without the mention of what we configure (even though it can be derived from context) a tad odd. I think a replacement like this would be preferable:
```suggestion
By moving the DNS server into LocalStack and configuring spawned AWS compute environments to use it by default, your Lambda functions, ECS containers, and EC2 instances should already be able to access LocalStack at `localhost.localstack.cloud`.
```

dfangl reviewed (commented):
I think it is a very nice blogpost! I just have some suggestions, of typos / sentences that seem kind of odd to me :)

simonrw committed:
Update content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md

Co-authored-by: Daniel Fangl <daniel.fangl@localstack.cloud>

simonrw committed:
Apply suggestions from code review

Co-authored-by: Daniel Fangl <daniel.fangl@localstack.cloud>

simonrw committed:
Add additional space

simonrw committed:
Add clarity around the certificate problem

simonrw committed:
Add line about binding to 0.0.0.0

simonrw committed:
Stupid typos 🤦

dfangl commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+    * if using host networking (`--network host`), requests are made to an IP address or `localhost`, which is not included in the certificate; and
+    * when using the gateway domain (`host.docker.internal`), this domain is also not included in the set of certificate domains.
+* Subdomains created by resources such as S3 buckets or OpenSearch clusters will not resolve to the LocalStack container.
+* Servers listening on the host must bind to either the Docker network gateway IP address, which is inconvenient to get, or `0.0.0.0` which exposes the server to traffic from outside the host, which may be insecure.
```
Is this sentence referring to containers running in a bridge network communicating with LS running with host networking?

dfangl commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/image-sketch.png:
I think the line around the docker logo seems out of place, especially since it is the only logo with a border

dfangl commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+    * **LocalStack presents a valid certificate for `*.localhost.localstack.cloud` domains**
+* Subdomains created by resources such as S3 buckets or OpenSearch clusters will not resolve to the LocalStack container.
+    * **Subdomains of `localhost.localstack.cloud` also resolve to the LocalStack container**. 
+* Each host port can only be bound once, whereas container ports are separate from each other and multiple containers can bind to the same port .
```
```suggestion
* Each host port can only be bound once, whereas container ports are separate from each other and multiple containers can bind to the same port.
```

dfangl reviewed (approved):
Looks very good now! Thanks for incorporating the feedback. I just have some minor questions/comments left, but it seems good to go from my side.

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+date: 2023-09-07
+lastmod: 2023-09-26
+tags:
+- news
```
I think it falls more under an engineering-focussed blog?

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+- Daniel Fangl
+---
+
+Connecting your application code to LocalStack is not always easy.
```
Can you maybe add a note why is that so (just a line or two)? This would give first-time readers some context about the problem that is being solved :)

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+## Connectivity to the LocalStack container
+
+LocalStack normally runs in a container, meaning that it is isolated from the host system.
+By default, LocalStack _publishes_ its edge port (usually 4566) to the host.
```
```suggestion
By default, LocalStack _publishes_ its edge port (usually `4566`) to the host.
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+Publishing a port means that a port on the host forwards network communications to the LocalStack container.
+Requests made to `localhost:4566` are then forwarded to the LocalStack container.
+
+This works well when interacting from the host, for example using `awslocal` commands.
```
```suggestion
This works well when interacting from the host, for example using [`awslocal`](https://docs.localstack.cloud/user-guide/integrations/aws-cli/#localstack-aws-cli-awslocal) commands.
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+
+This works well when interacting from the host, for example using `awslocal` commands.
+It does not work when trying to connect to LocalStack from your own containers, or LocalStack compute resources such as Lambda functions or ECS containers.
+In the past, some people have suggested connecting their application containers to the host network (`--network host`) or by making requests to `host.docker.internal:4566`.
```
```suggestion
In the past, our community users have suggested connecting their application containers to the host network (`--network host`) or by making requests to `host.docker.internal:4566`.
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
````diff
@@ -0,0 +1,184 @@
+dig @8.8.8.8 localhost.localstack.cloud
+```
+
+which outputs:
````
```suggestion
You will see the following output:
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+This command queries the Google public nameserver (`8.8.8.8`) for the `localhost.localstack.cloud` domain.
+In the "ANSWER" section we see `127.0.0.1` returned, as an `A` record, meaning IP address.
+
+### Why is using localhost.localstack.cloud not enough?
```
```suggestion
### Why is using `localhost.localstack.cloud` not enough?
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+
+### Why is using localhost.localstack.cloud not enough?
+
+When you create a lambda function, ECS container or EC2 instance, we create a new Docker container running your application code.
```
```suggestion
When you create a Lambda function, ECS container or EC2 instance, you create a new Docker container running your application code.
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+
+When you create a lambda function, ECS container or EC2 instance, we create a new Docker container running your application code.
+In these situations, using the domain name `localhost.localstack.cloud` will not resolve to the LocalStack container as you may expect.
+We currently have code in our Lambda managed runtimes to intercept any domain names that resolve to `127.0.0.1` and rewrite them to the LocalStack container IP, but this is only available to managed runtimes.
```
```suggestion
We currently have code in our Lambda managed runtimes to intercept any domain names that resolve to `127.0.0.1` and rewrite them to the LocalStack container IP. Currently this is only available to managed runtimes.
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+In these situations, using the domain name `localhost.localstack.cloud` will not resolve to the LocalStack container as you may expect.
+We currently have code in our Lambda managed runtimes to intercept any domain names that resolve to `127.0.0.1` and rewrite them to the LocalStack container IP, but this is only available to managed runtimes.
+
+To tackle the problem generally, we are bringing a feature from LocalStack Pro into LocalStack open source: our DNS server.
```
```suggestion
To tackle the problem generally, we are bringing our DNS server from LocalStack Pro into the LocalStack Community edition.
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+
+We are now able to resolve the three issues mentioned above:
+
+* If SSL is used, then certificate validation must be turned off since LocalStack does not present a valid certificate for the domain used(either `localhost` or `host.docker.internal`).
```
```suggestion
* If SSL is used, then certificate validation must be turned off since LocalStack does not present a valid certificate for the domain used (either `localhost` or `host.docker.internal`).
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+We are now able to resolve the three issues mentioned above:
+
+* If SSL is used, then certificate validation must be turned off since LocalStack does not present a valid certificate for the domain used(either `localhost` or `host.docker.internal`).
+    * **LocalStack presents a valid certificate for `*.localhost.localstack.cloud` domains**
```
```suggestion
    * LocalStack presents a valid certificate for `*.localhost.localstack.cloud` domains.
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+* If SSL is used, then certificate validation must be turned off since LocalStack does not present a valid certificate for the domain used(either `localhost` or `host.docker.internal`).
+    * **LocalStack presents a valid certificate for `*.localhost.localstack.cloud` domains**
+* Subdomains created by resources such as S3 buckets or OpenSearch clusters will not resolve to the LocalStack container.
+    * **Subdomains of `localhost.localstack.cloud` also resolve to the LocalStack container**. 
```
```suggestion
    * Subdomains of `localhost.localstack.cloud` also resolve to the LocalStack container. 
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+* Subdomains created by resources such as S3 buckets or OpenSearch clusters will not resolve to the LocalStack container.
+    * **Subdomains of `localhost.localstack.cloud` also resolve to the LocalStack container**. 
+* Each host port can only be bound once, whereas container ports are separate from each other and multiple containers can bind to the same port .
+    * **Now all networking can be done over the Docker network, and no ports have to be bound to the host at all**.
```
```suggestion
    * Now all networking can be done over the Docker network, and no ports have to be bound to the host at all.
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+### Setting LocalStack as the DNS server using Docker Compose
+
+When using Docker Compose, you can specify the IP address that the LocalStack container will be assigned by using a user-defined network, and using the `ipam` configuration settings.
+For example:
```
```suggestion
An example configuration would look similar to:
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
````diff
@@ -0,0 +1,184 @@
+        - subnet: 10.0.2.0/24
+```
+
+We have created a demo application to show off this functionality: [https://github.com/localstack/networking-demo-application](https://github.com/localstack/networking-demo-application).
````
```suggestion
We have created a demo application to show off this functionality: [github.com/localstack/networking-demo-application](https://github.com/localstack/networking-demo-application).
```

HarshCasper commented on content/blog/2023-09-07-making-connecting-to-localstack-easier/index.md:
```diff
@@ -0,0 +1,184 @@
+
+* The deploy process runs in a separate Docker container.
+* The application container connects across the Docker network to LocalStack.
+* A lambda function communicates with LocalStack to subscribe to SQS messages, access objects in S3, and write to a DynamoDB table.
```
```suggestion
* A Lambda function communicates with LocalStack to subscribe to SQS messages, access objects in S3, and write to a DynamoDB table.
```

simonrw committed:
Apply suggestions from code review

Co-authored-by: Harsh Mishra <erbeusgriffincasper@gmail.com>
Co-authored-by: Daniel Fangl <daniel.fangl@localstack.cloud>

simonrw committed:
Simplify usage examples

We should only really include one example, and we reduce the complexity
of the example.

SimonWallner commented:
PR continues in new repo: https://github.com/localstack/localstack-blog.github.io/pull/7
